### PR TITLE
Add codeGen plugin support

### PR DESCRIPTION
**Pull request by SerayaEryn**
This adds the possibility to replace `pug-code-gen` with a plugin.

### CODE REVIEW

File: packages/pug/lib/index.js [hunk 1]
```diff
@@ -50,6 +50,15 @@ function applyPlugins(value, options, plugins, name) {
   }, value);
 }
 
+function getPlugin(plugins, name, defaultFunction) {
+  for (var plugin of plugins) {
+    if (plugin[name]) {
+      return plugin[name];
+    }
+  }
+  return defaultFunction;
```
Comment (ForbesLindesay) [inline packages/pug/lib/index.js]: This function needs to throw an error if there is more than one plugin implementing the same `name`. i.e. this approach does not support multiple code-gen plugins.

Comment (SerayaEryn) [reply packages/pug/lib/index.js]: I decided to remove the new function and reuse `findReplacementFunc()` since it was almost doing what I needed.

**Comment by SerayaEryn** [comment]
That's totally fine. Thanks for the review. :)

File: packages/pug/lib/index.js [hunk 2]
```diff
@@ -172,7 +172,7 @@ function compileBody(str, options){
 
   // Compile
   ast = applyPlugins(ast, options, plugins, 'preCodeGen');
-  var js = generateCode(ast, {
+  var js = findReplacementFunc(plugins, 'codeGen', generateCode)(ast, {
```
Comment (ForbesLindesay) [inline packages/pug/lib/index.js]: If you made this:

```suggestion
  var js = (findReplacementFunc(plugins, 'codeGen') || generateCode)(ast, {
```

You wouldn't need any changes to `findReplacementFunc` at all.

Comment (SerayaEryn) [reply packages/pug/lib/index.js]: I made the suggested change and removed the changes to `findReplacementFunc()`.

**Pull request closed without merging**